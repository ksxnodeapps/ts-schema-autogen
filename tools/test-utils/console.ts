import { ConsoleInstance } from 'simple-fake-console'
import { Console } from '@ts-schema-autogen/types'

export class FakeConsole implements Console.Mod {
  public readonly core = new ConsoleInstance()
  public readonly info = jest.fn(this.core.info)
  public readonly error = jest.fn(this.core.error)
}
