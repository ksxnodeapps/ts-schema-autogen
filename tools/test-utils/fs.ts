import { StringPathFileSystem } from 'simple-fake-fs'
import { FSX } from '@ts-schema-autogen/types'

export class FakeFileSystem extends StringPathFileSystem implements FSX.Mod {
  private mkfn<Xs extends any[], Y> (fn: (...xs: Xs) => Y) {
    return jest.fn(async (...xs: Xs) => fn.apply(this, xs))
  }

  public readonly stat = this.mkfn(this.statSync)
  public readonly readdir = this.mkfn(this.readdirSync)
  public readonly readFile = this.mkfn(this.readFileSync)
  public readonly outputFile = this.mkfn(this.outputFileSync)
  public readonly writeFile = this.mkfn(this.writeFileSync)
  public readonly remove = jest.fn()
}
