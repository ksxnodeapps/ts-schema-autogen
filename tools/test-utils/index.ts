export * from './console'
export * from './fs'
export * from './path'
export * from './print-result'
export * from './process'
export * from './typescript-json-schema'
