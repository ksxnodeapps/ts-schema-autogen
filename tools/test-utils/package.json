{
  "name": "@tools/test-utils",
  "version": "0.0.0",
  "private": true,
  "dependencies": {
    "simple-fake-fs": "^0.0.8",
    "simple-fake-console": "^0.1.6",
    "simple-fake-path": "^0.0.2",
    "@ts-schema-autogen/status": "file:../../packages/status",
    "@ts-schema-autogen/types": "file:../../packages/types"
  }
}
