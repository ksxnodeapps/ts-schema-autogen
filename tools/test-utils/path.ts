import { Path } from '@ts-schema-autogen/types'

export class FakePath implements Path.Mod {
  constructor (private readonly cwd: string) {}

  public readonly join = (...paths: string[]): string => {
    if (!paths.length) return '.'
    const [head, ...rest] = paths
    const tail = this.join(...rest)
    if (head === '.') return tail
    if (tail === '.') return head
    if (tail.startsWith('../')) return this.join(this.dirname(head), tail.slice('../'.length))
    const left = head.endsWith('/') ? head.slice(0, -1) : head
    const right = tail.startsWith('/') ? tail.slice(1) : tail
    return left + '/' + right
  }

  public readonly resolve = (...paths: string[]): string => {
    if (!paths.length) return '.'
    const [head, ...rest] = paths
    const tail = this.resolve(...rest)
    if (tail.startsWith('/')) return tail
    return this.join(this.cwd, head, tail)
  }

  public readonly dirname = (path: string): string => {
    if (path === '' || path === '.' || path === '/') return path
    const segments = path.split('/').slice(0, -1)
    if (!segments.length) return '.'
    return this.join(...segments)
  }

  public readonly basename = (path: string): string => {
    const segments = path.split('/')
    return segments[segments.length - 1]
  }
}
