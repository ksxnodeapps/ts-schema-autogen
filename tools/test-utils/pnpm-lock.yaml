dependencies:
  '@ts-schema-autogen/types': 'link:../../packages/types'
  simple-fake-fs: 0.0.4
lockfileVersion: 5.1
packages:
  /@types/node/13.5.0:
    dev: false
    resolution:
      integrity: sha512-Onhn+z72D2O2Pb2ql2xukJ55rglumsVo1H6Fmyi8mlU9SvKdBk/pUSUAiBY/d9bAOF7VVWajX3sths/+g6ZiAQ==
  /simple-fake-fs/0.0.4:
    dependencies:
      '@types/node': 13.5.0
      tslib: 1.10.0
    dev: false
    resolution:
      integrity: sha512-GUwEIYXI6oSJdMZkP1hK3x084z6HX6uyHMrs3erJXEzdyMIrJ7HJEjlbtMrs6zsELvgx0wr7Y9zTK61+igHC8A==
  /tslib/1.10.0:
    dev: false
    resolution:
      integrity: sha512-qOebF53frne81cf0S9B41ByenJ3/IuH8yJKngAX35CmiZySA0khhkovshKK+jGCaMnVomla7gVlIcc3EvKPbTQ==
specifiers:
  '@ts-schema-autogen/types': ../../packages/types
  simple-fake-fs: ^0.0.4
