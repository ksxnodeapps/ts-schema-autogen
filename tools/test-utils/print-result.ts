import { inspect } from 'util'
import { Success, Failure } from '@ts-schema-autogen/status'

export function printFailure (result: Failure<unknown>): string {
  let text = '\n'
  result.print((...args) => text += args.join(' ') + '\n')
  return text
}

export const printResult = (result: Success<unknown> | Failure<unknown>): string =>
  result.code ? printFailure(result) : inspect(result)
