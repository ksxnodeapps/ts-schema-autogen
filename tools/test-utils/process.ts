import { Process } from '@ts-schema-autogen/types'

export class FakeProcess implements Process.Mod {
  constructor (private readonly workDir: string) {}
  public readonly cwd = jest.fn(() => this.workDir)
}
