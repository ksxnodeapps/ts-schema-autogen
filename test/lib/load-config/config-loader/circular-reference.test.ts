import fsTree from '../../../fixtures/fs-tree/circular-reference'
import { FakeFileSystem, FakePath, printResult } from '@tools/test-utils'

import {
  Status,
  CircularReference,
  ConfigLoader,
  createJsonConfigParser
} from '@ts-schema-autogen/lib'

async function setup (filename: string) {
  const fsx = new FakeFileSystem('/', fsTree)
  const path = new FakePath()
  const configLoader = new ConfigLoader({
    fsx,
    path,
    parsers: [
      await createJsonConfigParser('JSON Parser')
    ]
  })
  const result = await configLoader.loadConfig(filename)
  return { filename, fsx, path, configLoader, result }
}

describe('0 → 0', () => {
  const cfgFile = '1/0.json'

  it('result matches snapshot', async () => {
    const { result } = await setup(cfgFile)
    expect(result).toMatchSnapshot()
  })

  it('result has expected properties', async () => {
    const { result } = await setup(cfgFile)
    expect(result).toMatchObject({
      code: Status.CircularReference,
      error: expect.any(Array)
    })
  })

  it('result is a CircularReference', async () => {
    const { result } = await setup(cfgFile)
    expect(result).toBeInstanceOf(CircularReference)
  })

  it('error messages', async () => {
    const { result } = await setup(cfgFile)
    expect(printResult(result)).toMatchSnapshot()
  })

  it('status code', async () => {
    const { result } = await setup(cfgFile)
    expect(result.getStatusCode()).toBe(Status.CircularReference)
  })
})

describe('0 → 1 → 0', () => {
  const cfgFile = '2/0.json'

  it('result matches snapshot', async () => {
    const { result } = await setup(cfgFile)
    expect(result).toMatchSnapshot()
  })

  it('result has expected properties', async () => {
    const { result } = await setup(cfgFile)
    expect(result).toMatchObject({
      code: Status.CircularReference,
      error: expect.any(Array)
    })
  })

  it('result is a CircularReference', async () => {
    const { result } = await setup(cfgFile)
    expect(result).toBeInstanceOf(CircularReference)
  })

  it('error messages', async () => {
    const { result } = await setup(cfgFile)
    expect(printResult(result)).toMatchSnapshot()
  })

  it('status code', async () => {
    const { result } = await setup(cfgFile)
    expect(result.getStatusCode()).toBe(Status.CircularReference)
  })
})

describe('0 → 1 → 2 → 0', () => {
  const cfgFile = '3/0.json'

  it('result matches snapshot', async () => {
    const { result } = await setup(cfgFile)
    expect(result).toMatchSnapshot()
  })

  it('result has expected properties', async () => {
    const { result } = await setup(cfgFile)
    expect(result).toMatchObject({
      code: Status.CircularReference,
      error: expect.any(Array)
    })
  })

  it('result is a CircularReference', async () => {
    const { result } = await setup(cfgFile)
    expect(result).toBeInstanceOf(CircularReference)
  })

  it('error messages', async () => {
    const { result } = await setup(cfgFile)
    expect(printResult(result)).toMatchSnapshot()
  })

  it('status code', async () => {
    const { result } = await setup(cfgFile)
    expect(result.getStatusCode()).toBe(Status.CircularReference)
  })
})

describe('0 → (a0 → a1 → a0, b0 → b1 → b0)', () => {
  const cfgFile = 'tree/0.json'

  it('result matches snapshot', async () => {
    const { result } = await setup(cfgFile)
    expect(result).toMatchSnapshot()
  })

  it('result has expected properties', async () => {
    const { result } = await setup(cfgFile)
    expect(result).toMatchObject({
      code: Status.CircularReference,
      error: expect.any(Array)
    })
  })

  it('result is a CircularReference', async () => {
    const { result } = await setup(cfgFile)
    expect(result).toBeInstanceOf(CircularReference)
  })

  it('error messages', async () => {
    const { result } = await setup(cfgFile)
    expect(printResult(result)).toMatchSnapshot()
  })

  it('status code', async () => {
    const { result } = await setup(cfgFile)
    expect(result.getStatusCode()).toBe(Status.CircularReference)
  })
})

describe('0 → 1 → (a0 → a1 → 3 → 1, b0 → b1 → 3 → 1, c0 → c1 → 3 → 4)', () => {
  const cfgFile = 'tree/0.json'

  it('result matches snapshot', async () => {
    const { result } = await setup(cfgFile)
    expect(result).toMatchSnapshot()
  })

  it('result has expected properties', async () => {
    const { result } = await setup(cfgFile)
    expect(result).toMatchObject({
      code: Status.CircularReference,
      error: expect.any(Array)
    })
  })

  it('result is a CircularReference', async () => {
    const { result } = await setup(cfgFile)
    expect(result).toBeInstanceOf(CircularReference)
  })

  it('error messages', async () => {
    const { result } = await setup(cfgFile)
    expect(printResult(result)).toMatchSnapshot()
  })

  it('status code', async () => {
    const { result } = await setup(cfgFile)
    expect(result.getStatusCode()).toBe(Status.CircularReference)
  })
})
