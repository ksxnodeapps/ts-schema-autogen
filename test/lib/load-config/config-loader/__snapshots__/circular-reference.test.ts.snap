// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`0 → (a0 → a1 → a0, b0 → b1 → b0) error messages 1`] = `
"
CircularReference
    tree/0.json
    tree/1/a/0.json
    tree/1/a/1.json
"
`;

exports[`0 → (a0 → a1 → a0, b0 → b1 → b0) result matches snapshot 1`] = `
CircularReference {
  "code": 8,
  "error": Array [
    "tree/0.json",
    "tree/1/a/0.json",
    "tree/1/a/1.json",
  ],
  "value": undefined,
}
`;

exports[`0 → 0 error messages 1`] = `
"
CircularReference
    1/0.json
"
`;

exports[`0 → 0 result matches snapshot 1`] = `
CircularReference {
  "code": 8,
  "error": Array [
    "1/0.json",
  ],
  "value": undefined,
}
`;

exports[`0 → 1 → (a0 → a1 → 3 → 1, b0 → b1 → 3 → 1, c0 → c1 → 3 → 4) error messages 1`] = `
"
CircularReference
    tree/0.json
    tree/1/a/0.json
    tree/1/a/1.json
"
`;

exports[`0 → 1 → (a0 → a1 → 3 → 1, b0 → b1 → 3 → 1, c0 → c1 → 3 → 4) result matches snapshot 1`] = `
CircularReference {
  "code": 8,
  "error": Array [
    "tree/0.json",
    "tree/1/a/0.json",
    "tree/1/a/1.json",
  ],
  "value": undefined,
}
`;

exports[`0 → 1 → 0 error messages 1`] = `
"
CircularReference
    2/0.json
    2/1.json
"
`;

exports[`0 → 1 → 0 result matches snapshot 1`] = `
CircularReference {
  "code": 8,
  "error": Array [
    "2/0.json",
    "2/1.json",
  ],
  "value": undefined,
}
`;

exports[`0 → 1 → 2 → 0 error messages 1`] = `
"
CircularReference
    3/0.json
    3/1.json
    3/2.json
"
`;

exports[`0 → 1 → 2 → 0 result matches snapshot 1`] = `
CircularReference {
  "code": 8,
  "error": Array [
    "3/0.json",
    "3/1.json",
    "3/2.json",
  ],
  "value": undefined,
}
`;
